### DISCUSSION TITLE
Missing Waypoints in Routing

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#730] [Q&A]
I have a problem with the general routing in GLSP that the router is computing to less waypoints. 

This is an example of a BPMN model created with Eclipse BPMN2:

![image](https://user-images.githubusercontent.com/1339124/184597665-81e7482c-4ee6-44b1-a069-7c0faba38063.png)

This model contains an edge with 3 waypoints. The internal representation of the edge in BPMN 2.0 is:

```xml
      <bpmndi:BPMNEdge bpmnElement="SequenceFlow_2" id="BPMNEdge_SequenceFlow_2" sourceElement="BPMNShape_1" targetElement="BPMNShape_Task_1">
        <di:waypoint x="112.0" xsi:type="dc:Point" y="100.0"/>
        <di:waypoint x="112.0" xsi:type="dc:Point" y="76.0"/>
        <di:waypoint x="240.0" xsi:type="dc:Point" y="76.0"/>
        <bpmndi:BPMNLabel id="BPMNLabel_4"/>
      </bpmndi:BPMNEdge>
```

The model is rendered correctly with GLSP (OpenBPMN):

![image](https://user-images.githubusercontent.com/1339124/184597914-2d2a6d23-1460-40d0-892a-575a5df4d4e0.png)

And the model can also be rendered with other BPMN modellers like bpmn.io 

![image](https://user-images.githubusercontent.com/1339124/184598030-09cc6f54-a2ea-453f-ac20-2b32575c0fe9.png)

This all seems to be fine so far. 

But when I create a new edge with GLSP with Manhattan routing, the GLSP / Sprotty router is computing only 2 way points even if it looks like we have 4 way points:

![image](https://user-images.githubusercontent.com/1339124/184598825-3a30f42f-3edc-4d7a-b905-943ebc2a8255.png)

But internally there is only the start and end point computed. It looks like the GLSP Manhattan router is doing this on the fly. The internal representation contains only 2 waypoints:

```xml
    <bpmndi:BPMNEdge bpmnElement="SequenceFlow_RdCPXQ" id="BPMNEdge_yfemGg">
    	<di:waypoint x="112.0" y="118.0"/>
    	<di:waypoint x="258.0" y="267.0"/>
    </bpmndi:BPMNEdge>
```

As a result if you now open the diagram with Eclipse BPMN2 we have only a simple diagonal edge:

![image](https://user-images.githubusercontent.com/1339124/184599080-c66a4b6e-5845-4520-a9a7-475b160e43b3.png)

But as soon as you begin to drag the Edge in GLSP a little bit like this:

![Peek 2022-08-15 10-07](https://user-images.githubusercontent.com/1339124/184599516-24759aa1-8b02-479c-aa96-55e4de0499c2.gif)

the 4 waypoint are computed:

```xml
    <bpmndi:BPMNEdge bpmnElement="SequenceFlow_RdCPXQ" id="BPMNEdge_yfemGg">
    	<di:waypoint x="112.0" y="118.0"/>
    	<di:waypoint x="204.0" y="118.0"/>
    	<di:waypoint x="204.0" y="267.0"/>
    	<di:waypoint x="258.0" y="267.0"/>
    </bpmndi:BPMNEdge>
```

and of course the diagram is now displayed correctly in Eclipse BPMN or other BPMN modellers:

![image](https://user-images.githubusercontent.com/1339124/184599728-d7e26dd8-d9d7-4bcc-a9d8-0589fa2372c7.png)

Why is this happening and how can I force GLSP to compute all waypoints at the beginning? I do not really understand the internal behaviour of the router and why it shows up like Manhattan routing even if there is no Manhattan routing defined.

**Comment by tortmayr** [comment]
Hi @rsoika,
sorry for the late reply.

I had a look into this issue and at first glance to me this looks like a bug in the core GLSP framework.
TLDR: The issue is that we don't send any routing point information to the server when creating an new edge. So newly created edges will always have zero routing points on server side even if the client-side router computed additional points.

> But internally there is only the start and end point computed. It looks like the GLSP Manhattan router is doing this on the fly. The internal representation contains only 2 waypoints:

The GLSP Manhattan router is not doing it on the fly but the entire routing is computed on client side. This means currently the complete routing information is only available on the client side and we need to send it find a way to send it to the server as well.

>But as soon as you begin to drag the Edge in GLSP a little bit like this:

That's because the tool that handles moving (`ChangeBoundsTool`) has a [built in logic](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/tools/change-bounds-tool.ts#L250) to also update the routing points of  edges that are affected by the move.

Unfortunately we cannot reuse that same logic (i.e. the ``ChangeRoutingPointsOperation`) for the initial edge creation. Here I believe we have to update the `CreateEdgeOperation` and make it possible to also send the initial set of routing points, in addition to the source and target.

I have created https://github.com/eclipse-glsp/glsp/issues/738 to track this. We hopefully will be able to provide a fix soon.

**Comment by rsoika** [comment]
@tortmayr  thanks for your replay and the background information.
Is it possible that this topic is related to another issue I have in my pipeline:

Wenn you connect two elements, the default route is fine. Than you begin to move the element around in the diagram. Now the way points will stick on the diagram, even if the target anchor point is computed new depending on the current position.

But what I expect from such a modelling situation is, that the route is computed completely new when I start moving. This is how for example Eclipse-BPMN2  behaves. 

![routing-issue](https://user-images.githubusercontent.com/1339124/189080625-e3eaf3cd-df8d-43ea-a450-aae2e4106a29.gif)

If you want to change the layout in a large BPMN model to better explain a business situation, you typically move entire areas in the diagram. It makes no sense for the modelling to stick to the original waypoints after such an action. 

![routing-issue2](https://user-images.githubusercontent.com/1339124/189081809-e07440a8-3bf8-482d-a07d-954c06f673e9.gif)

So I'm thinking about the possibility of completely restarting the initial routing and dissolving existing waypoints.

**Comment by rsoika** [comment]
Is this topic somehow possible also related to https://github.com/eclipse-glsp/glsp/issues/612 ?
